### ISSUE TITLE
image sequence doesn't update in viewport render

### ISSUE THREAD

**Issue by neo2068** [LuxCoreRender/BlendLuxCore#318]
Windows 10 Pro, NVIDIA Geforce GTX 1080 + NVIDIA Geforce GTX 780
Blender version: 2.80
LuxCore version: 2.2

When I use an image sequence in imagemap node, the image didn't get updated in viewport rendering if current frame is changed.
The node has an update callback for the image pointer property and the image_user pointer property but they didn't get called after changing the frame.
To reproduce the issue, open the attached test scene, activate viewport rendering and change current frame in timeline.

[image sequence.zip](https://github.com/LuxCoreRender/BlendLuxCore/files/3776150/image.sequence.zip)

**Comment by Theverat** [comment]
The image sequence in your .blend file was set up incorrectly. 

I stumbled upon it because Cycles showed the same behaviour, then I checked the sequence settings, and the number of frames was set to 6 instead of 4 (I also had to fix the path/name because in the .blend, it was set to "new_luxcore_001.png", but the images were named "seq_001.jpg" etc.).

After I fixed the path and the number of frames, it works (in both Cycles and LuxCore).
[fixed.zip](https://github.com/LuxCoreRender/BlendLuxCore/files/3776384/fixed.zip)

**Comment by neo2068** [comment]
Sorry, then I packed the wrong .blend file in the zip archive. Your file works but when I try to fix it with myfile in luxcore only it didn't work. It only works if I set up the image texture in cycles additionally.

**Comment by Theverat** [comment]
Ah yes, it does not work without a Cyles image texture node indeed.

[image_sequence2.zip](https://github.com/LuxCoreRender/BlendLuxCore/files/3779397/image_sequence2.zip)

**Comment by Theverat** [comment]
Even with a Cycles texture node, the user has to enable "Auto Refresh" (node.image_user.use_auto_refresh).
Our custom nodes can't have the "official" image_user (it's a bpy_struct, just like color ramps and curves), so I'm not sure we can fix this.

**Comment by neo2068** [comment]
I have a workaround for the update problem. We can add a frame update handler in the export method of the imagemap class and update the image if there is a image sequence:

```
    def register_frame_handler(self):
        #Do not add multiple callbacks to handler stack.
        #Try to delete a previously added handler callback first and catch a potential value error if none exists
        #After that the callback can be added
        try:
            bpy.app.handlers.frame_change_pre.remove(self.frame_update)
        except ValueError:
            pass

        bpy.app.handlers.frame_change_pre.append(self.frame_update)

    def frame_update(self, scene):
        if self.image and self.image.source == "SEQUENCE":
            self.image = self.image

    def sub_export(self, exporter, depsgraph, props, luxcore_name=None, output_socket=None):
        [...]
        # Add frame update handler because image sequence doesn't get updated automatically
        self.register_frame_handler()

```
@Theverat What do you think?

**Comment by Theverat** [comment]
One issue I see is that it's hard to make sure that the handlers are removed when the node is deleted.

But the basic idea of using bpy.app.handlers is good, maybe we add a new file `frame_change_pre.py` here: https://github.com/LuxCoreRender/BlendLuxCore/tree/master/handlers
With something like this:

```python
from ..utils import node as utils_node

@persistent
def handler(scene):
    for node_tree in bpy.data.node_groups:
        for node in utils_node.find_nodes(node_tree, "LuxCoreNodeTexImagemap"):
            if node.image and node.image.source == "SEQUENCE":
                # Force a viewport update
                node.image = node.image
```

**Comment by neo2068** [comment]
Yes, that was an issue I also had in mind. Your propsal is a good solution. I will implement it and push it into master.

**Comment by Theverat** [comment]
I tested the code above and it looks like it will get a bit more nasty, because editing a node does not actually tag all it's users as updated. For example, try to run your or my code when the node tree is not opened in a node editor - it will not work. We have to flag the material as updated instead, but this leaves out other users of node trees, for example area lights.

**Comment by Theverat** [comment]
> I will implement it and push it into master.

I already have it almost working, so you don't have to do anything I think (just review my solution when I push it).
